### ISSUE TITLE
zrep not updating/setting src-host / dest-host on the receiving side

### ISSUE THREAD

**Issue by crispyduck00** [bolthole/zrep#187]
As mentioned in #186 zrep is not updating src-host / dest-host on the receiving host, to match sender / receiver.

Could be a issue after recovering from split brain e.g.:
```
host1: zrep -i tank/fs1 host2 tank/fs1
host1: zrep -S -f tank/fs1
host1: ifdown nic to host2
host2: zrep takeover tank/fs1 (fails)
host2: zrep takeover -L tank/fs1
host1: ifup nic to host2
host2: zrep -S -f tank/fs1 (fails; split brain)
host1: zrep -S -f tank/fs1 (fails; split brain)
host1: zfs inherit zrep:master tank/fs1
host2: zrep -S -f tank/fs1 (sync works again)
host1: zrep takeover tank/fs1 (fails)
host1: zfs get all tank/fs1 (dest and src host are host1)
```

Also if I manually change src and dest host on the receiving host and do afterwards a zrep -S or zrep -S -f  from the sender host, srxc and dest property is not updated. See:
```
host1: zrep -i tank/fs1 host2 tank/fs1
host1: zrep -S -f tank/fs1
host2: zfs get all tank/fs1 (zrep:src-host=host1 and zrep:dest-host=host2)
host2: zfs set zrep:src-host=xxx tank/fs1
host1: zrep -S -f tank/fs1
host1: zrep -S tank/fs1
host2: zfs get all tank/fs1 (zrep:src-host=xxx and zrep:dest-host=host2)
```

As described in #186 I added for testing `-o ${ZREPTAG}:src-host=$ZREP_SRC_HOST` to the sync part of the scripts zfs recv command, which will always update on the receiving side src-host to match the sending host.

Is there any case where this would not be applicable? I think src-host will always be the host from where the fs was last time received.

**Comment by ppbrown** [comment]
erm...
you are manually tweaking with the zrep properties with
"zfs inherit zrep:master tank/fs1"
does it say to do that in the docs?
If not, then I think this is a "you break it you bought it" case :-)

"proper" way to recover from this, when networking comes back, would be to
do a "failback", I think.
but thats just a quick off-the-cuff morning guess.
Please let me know if that doesnt work

**Comment by crispyduck00** [comment]
Not in the documentaition, but in the code (expire):
```
# Normally, dont quit on error. But this is super-bad.
	if [[ "$sanity" == "yes" ]] ; then
		zrep_errquit "Remote side also marked as master. to clear run on $remotehost: zfs inherit ${ZREPTAG}:master $remotefs"
	fi
```

Anyway, I dont want to failback, and even if so, I would first make the previously faulty host slave and be sure everything works and is sync again.
If needed then still a failover or as I preferre takeover can be done.

**Comment by ppbrown** [comment]
So.... your workflow isnt making sense to me.
to recap in english:
- you have working replication
- networking "fails"
- you force replication side to be master
- (  This implies you actually DO things with the data on the replication
side )
- then... networking comes back, and you force original to be master, and
.. what? throw away all data on replication side?
If you're going to do that, then making it master, almost doesnt make sense.

There are missing steps here.
based on what approach you want to take with the data, there is one of two
better paths that you should be taking.
Once I understand what you are doing, I will probably then change that
comment to be "go read the docs",
and write up a more lengthy section in the documentation.

**Comment by ppbrown** [comment]
oh wait. I was reading your steps wrong.
you are pushing the data from replication target, back to original master.

so basically, "zrep takeover" should be setting src and dest, but isnt.

**Comment by ppbrown** [comment]
Note to self:
prior logic flow just calls "ssh $remotehost zrep failover".
which presumes values are set as originally.
but in splitbrain scenario that wont be true.
So I need to add a "failover with explicit master " syntax.

**Comment by crispyduck00** [comment]
The network failure (ifdown/up) is just what I use for testing now.
In real it would maybe a psu failure, someone messing with the circuit breakers in the lab,...

So zrep master host1 has power outage, and due to my cluster hook scripts the zrep target/slave automatically becomes master (trying several times takeover and then forcing it).
From now on my second host2 acts as master serving my clients.
Now host1 one comes up again maybe uncontrolled, so both think they are master (split brain).
Now I want that the old master, host1 should be slave. Can be archived by inherit zrep master on it.
Now host2 is master and syncs to host1.
Great, everything looks good.

But on host1 the actual slave the zrep:src-host is still host1. Therefore a additional takeover back to host1 (normal or forced) fails.

Sure, can put it in my notes and never do zfs inherit master without changing src host at the same time. But think it is better to automate things. ;-)

Background:
I am running the whole thing on a Proxmox cluster having a LXC as NFS server expoting bind mounted zfs datasets.
The datasets are synced with zrep to the 2nd node. Zrep takeover is done by a LXC pre start hook script.
So I if I simple migrate the replicated LXC, the pre start script is doing normal takeover.
If normal takeover is not working it is performing a takeover -L.
Works really great.

**Comment by ppbrown** [comment]
okay... I added an explicit remote host option to failover, and takeover.
It seems like the right thing to do.
So in your situation, after you manually cleared master, you could
have called it, and it would clean up the other side.


however.... the real issue, is that you didnt realize there is
explicit documentation for your situation

http://www.bolthole.com/solaris/zrep/zrep.documentation.html#recovery1

 When you use failover, rather than takeover, there is no master/src confusion.

**Comment by crispyduck00** [comment]
Thx will check it.

I know failover works, but in my opinion if normal sync with -S or -S -f works, both, failover and takeover should work.

And primarily I don't want to do a failback at all, I just want that the old master becomes slave and everything is fine. 
I am just afraid that everything seems fine, sync works,... but after weeks a migration to the old master fails, as the takeover command fails because the host running the takeover still has the src-host prop pointing to itself.

Maybe my english description is not making the use case and potential overal issues clear.

**Comment by ppbrown** [comment]
Note to myself: check to see that all versions of takeover and failover,
set BOTH  src and dest hosts.

**Comment by ppbrown** [comment]
Note to C.D.:
"I don't want to do a failback at all, I just want that the old master
becomes slave and everything is fine"

lol.. I think you are getting hung up on the word "fail".
What you describe is pretty much what failover does.

**Comment by crispyduck00** [comment]
Hmm, seems I have to check this again. For me it was always failover is the same as takeover, but initiated from the sending side.

So if I want to switch master from host 1 to 2 I can either do takeover on host2 or I can do a failover on host1.

Maybe I got here something wrong.

From the doc:
```
If you had to force the secondary to be master (via "takeover") due to primary server being down, then you need to first bring the pair to a partially synced state again, by rolling back any changes since the last sync, on the master. Then you can do another failover command on the secondary.
All steps required are shown below:

	host1# zrep list -v pool1/prodfs
         # Output will show last synced snapshot

    	host1# zfs rollback (last snapshot) pool1/prodfs

    	host2# zfs failover pool1/prodfs   
```
Who is master after this? In my understanding this will be host1, and that is not what I want.
Host2 should stay master and host2 should be accepted as src.
Wrong?

**Comment by ppbrown** [comment]
In the example in the docs:
host1 used to be master.  host2 is currently LIVE master. host one thinks
it is, but isnt really production master.

Following the example makes host1 true master master again.

This is literally what you described you wanted in the zrep issue, as I
understood it.
You started with your host1 as master.
You faked a "network outage."
you forced host2 to be master.
But once network is fixed, you want host1 to go back to being real master.

**Comment by crispyduck00** [comment]
> You started with your host1 as master. You faked a "network outage." you forced host2 to be master. But once network is fixed, you want host1 to go back to being real master.

No, host1 should become slave after split brain. And should be ready for a eventually takeover.

**Comment by ppbrown** [comment]
Ah. your intention was not clear in the initial example you gave.

And funily enough, thinking about your needs,  made me look at the current code.. which gave me a way to improve the docs a lot as well, from when they were originally written.

It would be nice if you could verify the following as valid steps:

   ===
   
(host1= original master before split, host2 is master currently)

 host1# zrep failover -L pool1/prodfs
       (this will do a rollback on host1, but not touch host2)

host2# zrep -S pool1/prodfs

Now both sides are synced, and  it is your choice whether to do failback from host2, or takeover from host1.

**Comment by crispyduck00** [comment]
> host1= original master before split, host2 is master currently)
> 
> host1# zrep failover -L pool1/prodfs (this will do a rollback on host1, but not touch host2)
> 
> host2# zrep -S pool1/prodfs

Yes this works now.
Think it will also have working in the past when using the failover -L, important is the -L.

Without the -L and before adding the remotemaster check (#181) doing a failover would have overwritten any changes made on host2.

This was also the main reason I did not use/try it in this case with failover.

So yes, failover -L is exactly doing what is needed here, doing local rollback, updating src/dest host and removing the master property. 
But, do we really want to use a function for this that is usually, (without the -L parameter) doing a local snapshot that is sent over to the remote host? 
Maybe I am just paranoid, but this was the main reason not trying/testing it with failover. 

The most important in this scenario is that the actual master in this case host2 is not touched, what is needed is a eventual rollback on host1, correct or updated src/dest on host1 and remove of master on host1, failover -L is doing all this. 
But I am still not comfortable with it using here a function that usually without -L will snap and sync over the actual fs before doing the master switch. What it also did before #181.

Tested it several times, with many different scenarios, writing data to host1 and 2 in between,... and it works correctly and the remotemaster/splitbrian check is also preventing any harm if the -L is not used. 
So can be closed, if the above mentioned is only my personal paranoia/feeling with it. ;-)
